src/adagrams.js

```diff
@@ -1,15 +1,158 @@
+      if (poolOfLettersDict[choice] > 0) {
+        lettersInHand.push(choice);
+        counter += 1;
+        poolOfLettersDict[choice] -= 1;
+      }
+    }
+    return lettersInHand;
+  };
 
-export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
-};
+  usesAvailableLetters = (input, lettersInHand) => {
+    const dictLettersInHand = {};
+    for (let i = 0; i < lettersInHand.length; i++) {
+      if (lettersInHand[i] in dictLettersInHand) {
+        let val = dictLettersInHand[lettersInHand[i]];
+        dictLettersInHand[lettersInHand[i]] = val + 1;
+      } else {
+        dictLettersInHand[lettersInHand[i]] = 1;
+      }
+    }
+    for (let i = 0; i < input.length; i++) {
+      if (input[i] in dictLettersInHand) {
+        let val1 = dictLettersInHand[input[i]];
+        val1 = val1 - 1;
+        dictLettersInHand[input[i]] = val1;
+        if (val1 === 0) {
+          delete dictLettersInHand[input[i]];
+        }
+      } else {
+        return false;
+      }
+    }
+    return true;
+  };
 
-export const scoreWord = (word) => {
-  // Implement this method for wave 3
-};
+  scoreWord = (word) => {
+    // Implement this method for wave 3
+    let scoreDict = {
+      A: 1,
+      B: 3,
+      C: 3,
+      D: 2,
+      E: 1,
+      F: 4,
+      G: 2,
+      H: 4,
+      I: 1,
+      J: 8,
+      K: 5,
+      L: 1,
+      M: 3,
+      N: 1,
+      O: 1,
+      P: 3,
+      Q: 10,
+      R: 1,
+      S: 1,
+      T: 1,
+      U: 1,
+      V: 4,
+      W: 4,
+      X: 8,
+      Y: 4,
+      Z: 10,
+    };
 
-export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
-};
+    let score = 0;
+    word = word.toUpperCase();
+    for (let i = 0; i < word.length; i++) {
+      if (word.charAt(i) in scoreDict) {
+        score = score + scoreDict[word.charAt(i)];
+      }
+    }
+    if (word.length > 6) {
+      score += 8;
+    }
+    return score;
+  };
+  // pseudocode for wave 4:
+  // Call the above scoreWord function and calculate score of each word within the word array and
+  // store it in a dictionary object named wordScoreDict with key as the word and value as score of each word and
+  // in the same loop keep a track of the maxscore
+  // initialize another resultant dictionary winningWord which will return the winning word and score
+  // Loop through all keys in wordScoreDict dictionary object and check if any value matches with
+  // maxvalue. If it does match, checjk if the length of that word is equal to 10.
+  // if it equals 10, then return it and break the loop
+  // also check minimum word length side by side in each key value and at the end, return the word
+  // which has min word length and its score in result dictionary object.
+
+  highestScoreFrom = (words) => {
+    // Implement this method for wave 4
+    let wordScoreDict = {};
+    let maxScore = -1;
+    let maxWord = "";
+    for (let i = 0; i < words.length; i++) {
```

Comment:
This would be a good place to use `for word in words`, similar to what you would do in Python. That way instead of using `words[i]`, you can just use `word`. It makes the code look cleaner.